### PR TITLE
Feature/cc 117 dashboard search user account

### PR DESCRIPTION
**Pull request by PengchengYuGithub**
This card enables searching the user by three specific fields: user's ID/ email/ name, or all of them. User's ID search requires exact match, but email/ name search only need to include the keyword. Name search also allows the expression of "firstName lastName" with a space to split them.

### CODE REVIEW

File: src/users/user.service.ts [hunk 1]
```diff
@@ -40,6 +41,59 @@ export class UserService {
...
+            })
+          : (optionalQuery = {
+              $or: [{ firstName: qName }, { lastName: qName }, { email: qEmail }],
+            });
+      }
+    }
+    if (user_id && !email && !name) {
+      user_id.match(/^[0-9a-fA-F]{24}$/)
+        ? (optionalQuery = { _id: user_id })
+        : (optionalQuery = { email: user_id + "@wrong" });
```
Comment (Double-Twelve) [inline src/users/user.service.ts:80]: what does this mean?

File: src/users/user.controller.ts [hunk 2]
```diff
@@ -31,11 +32,11 @@ export class UserController {
...
   @Get()
-  async listAllUsers(@Query() paginationQuery: PaginationQueryDto): Promise<{
-    total: number;
-    data: User[];
-  }> {
-    return await this.userService.getAllUsers(paginationQuery);
+  async listUsersByAdmin(
+    @Query() searchUserDto: SearchUserDto,
+  ): Promise<{ data: User[]; total: number }> {
+    console.log(searchUserDto);
```
Comment (Double-Twelve) [inline src/users/user.controller.ts:38]: please remove it

File: src/users/user.controller.ts [hunk 3]
```diff
@@ -31,11 +32,11 @@ export class UserController {
...
+    return await this.userService.getUserBySearch(searchUserDto);
```
Comment (Double-Twelve) [inline src/users/user.controller.ts:39]: if don't use _getAllUsers_, please delete it in the corresponding service file.

File: src/users/user.service.ts [hunk 4]
```diff
@@ -40,6 +41,59 @@ export class UserService {
     return { data: users, total };
   }
 
+  async getUserBySearch(searchUserDto: SearchUserDto): Promise<{ data: User[]; total: number }> {
+    let users = [];
+    let optionalQuery = {};
+    let splitName: string[];
+    let qFirstName, qLastName, qName, qEmail: RegExp;
```
Comment (Double-Twelve) [inline src/users/user.service.ts:48]: why define these variables in advance, except _optionalQuery_

File: src/users/user.service.ts [hunk 5]
```diff
@@ -40,6 +41,59 @@ export class UserService {
...
+    const { user_id, email, name, limit = 0, offset = 0, sort, desc } = searchUserDto;
+    const sorting = sort ? { [sort]: desc } : { createdAt: -1 };
+    // eslint-disable-next-line prefer-const
+    qEmail = new RegExp(`.*${email}.*`, "i");
+    if (name && name.includes(" ")) {
+      splitName = name.split(" ");
+      qFirstName = new RegExp(`.*${splitName[0]}.*`, "i");
+      qLastName = new RegExp(`.*${splitName[1]}.*`, "i");
```
Comment (Double-Twelve) [inline src/users/user.service.ts:56]: if name include multiple spaces, how to solve it?

**Comment by Double-Twelve** [comment]
Please write tests for your newly added methods

File: src/users/user.service.ts (hunk 5 above)
Comment (PengchengYuGithub) [reply src/users/user.service.ts:56]: In registration process of accounts, space is forbidden in firstName and lastName (one card was created before to do this); thus it should not include extra space in the name.

File: src/users/user.service.ts [hunk 6]
```diff
@@ -40,6 +41,59 @@ export class UserService {
...
+    if (user_id && !email && !name) {
+      user_id.match(/^[0-9a-fA-F]{24}$/)
+        ? (optionalQuery = { _id: user_id })
+        : (optionalQuery = { email: user_id + "@wrong" });
+    }
+    if (email && !user_id && !name) {
+      optionalQuery = { email: qEmail };
+    }
+    users = await this.userModel
+      .find(optionalQuery, { isDeleted: false })
```
[the remark this answers is not in the document]
Comment (PengchengYuGithub) [reply src/users/user.service.ts:86]: Corrected. Thx for the careful review.

File: src/users/user.service.ts (hunk 1 above)
Comment (PengchengYuGithub) [reply src/users/user.service.ts:80]: Different from "FindOne" or "FindById" method in MongoDb, if I do not "Find" the _id field by the string of a standard 12-byte ObjectId, there would be a 500 error. So I just search a safe text that would return empty result before. Now I would just return out of the function when user_id is not 12-byte string.

File: src/users/user.service.ts (hunk 4 above)
Comment (PengchengYuGithub) [reply src/users/user.service.ts:48]: Corrected.

**Pull request closed without merging**